girder/cli/shell.py

```diff
@@ -42,15 +44,24 @@ def _launchShell(context):
         return code.interact(banner=header, local=context)
 
 
-@click.command('shell', short_help='Run a Girder shell.', help='Run a Girder shell.')
+@click.command('shell', short_help='Run a Girder shell.', help='Run an interactive Girder shell '
+               'or a script in the Girder environment.')
 @click.option('--plugins', default=None, help='Comma separated list of plugins to import.')
-def main(plugins):
+@click.argument('script', type=click.Path(exists=True, dir_okay=False), required=False)
+@click.argument('args', nargs=-1, required=False)
+def main(plugins, script, args):
     if plugins is not None:
         plugins = plugins.split(',')
 
     webroot, appconf = configureServer(plugins=plugins)
 
-    _launchShell({
-        'webroot': webroot,
-        'appconf': appconf
-    })
+    if script is None:
+        _launchShell({
+            'webroot': webroot,
+            'appconf': appconf
+        })
+    else:
+        globals_ = {k: v for k, v in six.viewitems(globals()) if k != '__name__'}
```

Comment:
Why does every global have to be involved?

Reply:
To exec a file in a similar manner as `python <file>`, you have to pass the globals.

Pedantically, this is fine *but* won't work for a file that uses the multiprocessing module.  Multiprocessing is a special case where you actually have to mutate the `globals()` object rather than copy it to a new dictionary.

For that purpose, this could be
```
    exec_ = six.exec_
    globals_ = globals()
    originalGlobals = globals_.copy()
    # if we don't want the receiving file to already have our imports, we may want to remove any key from globals_ that doesn't start with _
    globals_.update({
        '__name__': '__main__', 
        '__file__': script,
        'webroot': webroot,
        'appconf': appconf,
    })
    exec_(open(script, 'rb').read(), globals_)
    globals._clear()
    globals_.update(originalGlobals)

Reply:
See https://github.com/manthey/pyexe/blob/master/pyexe.py#L76-L114 for even more thorough support (python files may be zip archives, for instance).

Reply:
> Pedantically, this is fine but won't work for a file that uses the multiprocessing module. Multiprocessing is a special case where you actually have to mutate the globals() object rather than copy it to a new dictionary.

Is this documented somewhere? It seems like a bug

Reply:
I think it has to do with this code https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Lib/multiprocessing/__init__.py#L22.  A contributor to the Octave project made me aware of it.  I wasn't able to find  an issue in the Python issue tracker (but there very well could be one).